reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,120 @@
+import java.util.List;
+import java.util.Optional;
 
 public class LiteralToUniformReductionOpportunity
     extends AbstractReductionOpportunity {
 
-  private final Expr literalExpr;
-  private final PipelineInfo pipelineInfo;
+  private final ConstantExpr literalExpr;
+  private final ShaderJob shaderJob;
 
-  LiteralToUniformReductionOpportunity(Expr literalExpr,
-                                       PipelineInfo pipelineInfo,
+
+  LiteralToUniformReductionOpportunity(ConstantExpr literalExpr,
+                                       ShaderJob shaderJob,
                                        VisitationDepth depth) {
+
     super(depth);
+
     this.literalExpr = literalExpr;
-    this.pipelineInfo = pipelineInfo;
+    this.shaderJob = shaderJob;
+
   }
 
   @Override
   void applyReductionImpl() {
-    // TODO: ...
+
+    final String arrayName = Constants.INT_LITERAL_UNIFORM_VALUES;
+    final List<Number> values;
+    final int index;
+
+    // Creates a combined list of the current and previously added values.
+    // Updates the index number to point at the corresponding element in the array.
+    if (this.shaderJob.getPipelineInfo().hasUniform(arrayName)) {
+      values = this.shaderJob.getPipelineInfo().getArgs(arrayName);
+      if (values.contains(((IntConstantExpr)this.literalExpr).getNumericValue())) {
+        index = values.indexOf(((IntConstantExpr)this.literalExpr).getNumericValue());
+      } else {
+        index = this.shaderJob.getPipelineInfo().getArgs(arrayName).size();
+        values.add(((IntConstantExpr)this.literalExpr).getNumericValue());
+      }
+
+    } else {
+      values = new ArrayList<>();
+      values.add(((IntConstantExpr)this.literalExpr).getNumericValue());
+      index = values.size() - 1;
+    }
+
+    // Removes the old uniform array and adds the new one to the associated json structure.
+    this.shaderJob.getPipelineInfo().removeUniform(arrayName);
+    final int binding = this.shaderJob.getPipelineInfo().getNumUniforms();
+    this.shaderJob.getPipelineInfo().addUniform(arrayName, BasicType.INT,
+        Optional.of(values.size()), values);
+    this.shaderJob.getPipelineInfo().addUniformBinding(arrayName, false, binding);
+
+    // Adds declaration for the uniform array and replaces the literal with an element in the array.
+    for (TranslationUnit tu: shaderJob.getShaders()) {
+
+      // Declares the array if it didn't exist. Otherwise, removes it and adds a new declaration
+      // with increased size.
+      Optional<Declaration> oldDeclaration = Optional.empty();
+      final List<Declaration> declarations = tu.getTopLevelDeclarations();
+      for (Declaration declaration : declarations) {
+        if (declaration instanceof InterfaceBlock) {
+          final List<String> members = ((InterfaceBlock) declaration).getMemberNames();
+          if (members.get(0).equals(arrayName)) {
+            oldDeclaration = Optional.of(declaration);
+          }
+        }
```

Comment:
We should apply this transformation to shaders where uniforms are not in interface blocks - so just look out for a global uniform variable with the right name.

Reply:
Fixed.